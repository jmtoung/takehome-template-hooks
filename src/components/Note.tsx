import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { Button, Row, Image, Col } from 'react-bootstrap';
import { useInput } from '../hooks/useInput';
import { API } from '../api';
import { NoteForm } from './NoteForm';

const deleteIcon = require('../assets/delete.png');
const editIcon = require('../assets/pencil.png');

export const Note: React.FunctionComponent<{
  id: string;
  title: string;
  body: string;
  deleteNote: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}> = (props) => {
  const [isUpdating, setIsUpdating] = useState(false);
  const { id } = props;
  const { value: title, onChange: changeTitle } = useInput(props.title);
  const { value: body, onChange: changeBody } = useInput(props.body);

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    await API.updateNote(id, title, body);
    setIsUpdating(false);
  };

  let view;

  if (isUpdating) {
    view = (
      <NoteForm
        text="Update Note"
        title={title}
        body={body}
        handleSubmit={(e) => handleSubmit(e)}
        handleBodyChange={changeBody}
        handleTitleChange={changeTitle}
      />
    );
  } else {
    view = (
      <div>
        <Row>
          <Col>
            <Link to={{ pathname: `/note/${id}` }}>
              <h2>{title}</h2>
            </Link>
          </Col>
          <Button
            type="button"
            style={{ margin: '5px' }}
            className="update-view"
            onClick={() => setIsUpdating(true)}
          >
            <Image alt="edit" src={editIcon} width="20" height="20" />
          </Button>
          <Button
            style={{ margin: '5px' }}
            type="button"
            className="delete"
            data-id={id}
            variant="secondary"
            onClick={props.deleteNote}
          >
            <Image alt="delete" src={deleteIcon} width="20" height="20" />
          </Button>
        </Row>
        <Row>
          <Col>
            <p>{body}</p>
          </Col>
        </Row>
      </div>
    );
  }

  return <div id={id}>{view}</div>;
};
