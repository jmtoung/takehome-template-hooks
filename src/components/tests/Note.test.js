import React from 'react';
import Note from '../Note';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

describe('Note', () => {
  it('renders note correctly.', () => {
    const wrapper = shallow(<Note />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('calls delete function when delete button is clicked.', () => {
    const deleteFunc = jest.fn();
    const wrapper = shallow(<Note deleteNote={deleteFunc} />);
    wrapper
      .find('.delete')
      .simulate('click');
    expect(deleteFunc).toHaveBeenCalledTimes(1);
  });

  it('calls updateView function and view is updated.', () => {
    const wrapper = shallow(<Note
      id="1"
      title='This is a title'
      body='This is the body.'
    />);
    const updateView = jest.spyOn(wrapper.instance(), 'updateView');

    expect(wrapper.state()).toEqual({ // State of component before click.
      id: '1',
      title: 'This is a title',
      body: 'This is the body.',
      isUpdating: false,
    });

    wrapper.find('.update-view').simulate('click');

    expect(updateView).toHaveBeenCalledTimes(1);
    expect(wrapper.state()).toEqual({ // State of component after click.
      id: '1',
      title: 'This is a title',
      body: 'This is the body.',
      isUpdating: true,
    });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('calls on api to update note when submit is called.', (done) => {
    const mockSuccessResponse = {id: '1', title: 'This is the updated title', body: 'This is the updated body.'};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockFetchPromise);

    const wrapper = shallow(<Note id='1'/>);
    var submitEvent = new Event('submit');
    wrapper.instance().handleSubmit(submitEvent);


    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith("https://5ec581652a4ba000163d2084.mockapi.io/api/v1/notes/1", {"body": "{}", "headers": {"Content-Type": "application/json"}, "method": "PUT"});

    process.nextTick(() => {
      expect(wrapper.state()).toEqual({
        id: '1',
        title: 'This is the updated title',
        body: 'This is the updated body.',
        isUpdating: true
      });
    });

    global.fetch.mockClear();
    done();
  });

});