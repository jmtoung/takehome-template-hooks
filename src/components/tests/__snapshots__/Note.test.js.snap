// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note calls updateView function and view is updated. 1`] = `
<div
  id="1"
>
  <div
    onClick={[Function]}
  >
    <NoteForm
      body="This is the body."
      handleChange={[Function]}
      handleSubmit={[Function]}
      text="Update Note"
      title="This is a title"
    />
  </div>
</div>
`;

exports[`Note renders note correctly. 1`] = `
<div>
  <div>
    <Row
      noGutters={false}
    >
      <Col>
        <Link
          from="/undefined"
          to={
            Object {
              "pathname": "/note/undefined",
              "state": Object {
                "body": undefined,
                "title": undefined,
              },
            }
          }
        >
          <h2 />
        </Link>
      </Col>
      <Button
        active={false}
        className="update-view"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "margin": "5px",
          }
        }
        type="button"
        variant="primary"
      >
        <Image
          alt="edit"
          fluid={false}
          height="20"
          rounded={false}
          roundedCircle={false}
          src={
            Object {
              "default": "pencil.png",
            }
          }
          thumbnail={false}
          width="20"
        />
      </Button>
      <Button
        active={false}
        className="delete"
        disabled={false}
        style={
          Object {
            "margin": "5px",
          }
        }
        type="button"
        variant="secondary"
      >
        <Image
          alt="delete"
          fluid={false}
          height="20"
          rounded={false}
          roundedCircle={false}
          src={
            Object {
              "default": "delete.png",
            }
          }
          thumbnail={false}
          width="20"
        />
      </Button>
    </Row>
    <Row
      noGutters={false}
    >
      <Col>
        <p />
      </Col>
    </Row>
  </div>
</div>
`;
